### ISSUE TITLE
[SPIR-V] generated SPIR-V is invalid: ID 'XXX[%XXX]' has not been defined

### ISSUE THREAD

**Issue by fnuecke** [microsoft/DirectXShaderCompiler#7025]
**Description**
I ran into an unexpected fatal error when compiling a mesh shader with debug symbols.  Compilation passes successfully without debug symbols enabled (i.e. without the `-fspv-debug=vulkan-with-source` flag).

**Steps to Reproduce**
Repro on godbolt: https://godbolt.org/z/oh8fjYcb8

Given this shader code:
```hlsl
// test.hlsl

#define MAX_VERTICES 64
#define MAX_PRIMITIVES 124

struct MeshPayload {
  uint data[32];
};

struct VertexAttributes {
    float3 position : SV_Position;
};

[outputtopology("triangle")]
[numthreads(1, 1, 1)]
void main(
  in payload MeshPayload payload,
  uint gid : SV_GroupID,
  out indices uint3 triangles[MAX_PRIMITIVES],
  out vertices VertexAttributes vertices[MAX_VERTICES]
) {
  uint data = payload.data[gid];
}
```

Run DXC with the following command:

```sh
dxc -T ms_6_5 -spirv -fspv-target-env=vulkan1.1spirv1.4 -fspv-debug=vulkan-with-source test.hlsl
```

**Actual Behavior**

Output:
```sh
fatal error: generated SPIR-V is invalid: ID '356[%356]' has not been defined
  %355 = OpExtInst %void %1 DebugValue %71 %356 %75 %int_0
```

Expected: no error, debug information embedded in SPIR-V output.

**Environment**
- DXC version:
  - July 2024: `libdxcompiler.so: 1.8(dev;4662-416fab6b); libdxil.so: 1.8`
  - Trunk: `libdxcompiler.so: 1.8(dev;4773-eb1223e0); libdxil.so: 1.8`
- Host Operating System
  - Manjaro Linux
  - uname -a: `Linux guybrush 6.11.10-2-MANJARO #1 SMP PREEMPT_DYNAMIC Mon, 25 Nov 2024 05:29:44 +0000 x86_64 GNU/Linux`

----------

I'm aware that resources are limited and debug-related issues are not a priority.  If you can provide me with some pointers on where to start, I might have a look if I can figure out some more details.  Thank you!

**Comment by Keenuts** [comment]
Thanks!

This looks to be caused by the `copy-propagate-arrays` pass during the legalization step.
The DebugValue was previously generated because a local variable was modified.
The pass created a new `OpAccessChain` directly accessing the array without going through the local variable (and will let DCE remove this dead code), and patched the `DebugValue` instruction.

Issue is the `OpAccessChain` is added after the patched `DebugValue` instruction, causing an invalid forward reference.
